test/unit/org/apache/cassandra/db/streaming/EntireSSTableStreamConcurrentComponentMutationTest.java

```diff
@@ -0,0 +1,338 @@
+
+    @BeforeClass
+    public static void defineSchemaAndPrepareSSTable()
+    {
+        SchemaLoader.prepareServer();
+        SchemaLoader.createKeyspace(KEYSPACE,
+                                    KeyspaceParams.simple(1),
+                                    SchemaLoader.standardCFMD(KEYSPACE, CF_STANDARD),
+                                    SchemaLoader.compositeIndexCFMD(KEYSPACE, CF_INDEXED, true),
+                                    SchemaLoader.standardCFMD(KEYSPACE, CF_STANDARDLOWINDEXINTERVAL)
+                                                .minIndexInterval(8)
+                                                .maxIndexInterval(256)
+                                                .caching(CachingParams.CACHE_NOTHING));
+
+        Keyspace keyspace = Keyspace.open(KEYSPACE);
+        store = keyspace.getColumnFamilyStore("Standard1");
+
+        // insert data and compact to a single sstable
+        CompactionManager.instance.disableAutoCompaction();
+        for (int j = 0; j < 10; j++)
+        {
+            new RowUpdateBuilder(store.metadata(), j, String.valueOf(j))
+            .clustering("0")
+            .add("val", ByteBufferUtil.EMPTY_BYTE_BUFFER)
+            .build()
+            .applyUnsafe();
+        }
+        store.forceBlockingFlush();
+        CompactionManager.instance.performMaximal(store, false);
+
+        sstable = store.getLiveSSTables().iterator().next();
+        descriptor = sstable.descriptor;
+
+        Token start = ByteOrderedPartitioner.instance.getTokenFactory().fromString(Long.toHexString(0));
+        Token end = ByteOrderedPartitioner.instance.getTokenFactory().fromString(Long.toHexString(100));
+        rangesAtEndpoint = RangesAtEndpoint.toDummyList(Collections.singleton(new Range<>(start, end)));
+
+        service = Executors.newFixedThreadPool(2);
+    }
+
+    @AfterClass
+    public static void cleanup()
+    {
+        service.shutdown();
+    }
+
+    @After
+    public void reset() throws IOException
+    {
+        latch = new CountDownLatch(1);
+        // reset repair info to avoid test interfering each other
+        descriptor.getMetadataSerializer().mutateRepairMetadata(descriptor, 0, ActiveRepairService.NO_PENDING_REPAIR, false);
+    }
+
+    @Test
+    public void testStream() throws Exception
+    {
+        testStreamWithConcurrentComponentMutation(NO_OP, NO_OP);
+    }
+
+    /**
+     * Entire-sstable-streaming receiver will throw checksum validation failure because concurrent stats metadata
+     * update causes the actual transfered file size to be different from the one in {@link ComponentManifest}
+     */
+    @Test
+    @BMRule(name = "Let component mutation happen immediately after ComponentManifest completion",
+    targetClass = "ComponentManifest",
+    targetMethod = "<init>",
+    action = "Thread.sleep(3000)")
+    public void testStreamWithStatsMutation() throws Exception
+    {
+        testStreamWithConcurrentComponentMutation(NO_OP, () -> {
```

Comment:
@jasonstack Crazy idea: What if we just did the mutation as the `runBeforeStreaming` callable here? We could at least avoid the hard-coded sleep (although the `mutation` callable might need a more generalized name, like `runConcurrentWithStreaming` or something).

Reply:
+1, this will prove the need of on-demand manifest creation, rather than relying on the manifest created initially.